### ISSUE TITLE
_ClusterTaggableManager breaking manage.py update_index

### ISSUE THREAD

**Issue by KaitCrawford** [torchbox/django-modelcluster#54]
When running manage.py update_index we are encountering this error:

```
Traceback (most recent call last):
  File "/var/praekelt/molo-tuneme/manage.py", line 11, in <module>
    execute_from_command_line(sys.argv)
  File "/var/praekelt/python/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/var/praekelt/python/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/var/praekelt/python/local/lib/python2.7/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/var/praekelt/python/local/lib/python2.7/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/var/praekelt/python/local/lib/python2.7/site-packages/wagtail/wagtailsearch/management/commands/update_index.py", line 87, in handle
    self.update_backend(backend_name, object_list)
  File "/var/praekelt/python/local/lib/python2.7/site-packages/wagtail/wagtailsearch/management/commands/update_index.py", line 48, in update_backend
    for chunk in self.print_iter_progress(self.queryset_chunks(queryset)):
  File "/var/praekelt/python/local/lib/python2.7/site-packages/wagtail/wagtailsearch/management/commands/update_index.py", line 103, in print_iter_progress
    for i, value in enumerate(iterable, start=1):
  File "/var/praekelt/python/local/lib/python2.7/site-packages/wagtail/wagtailsearch/management/commands/update_index.py", line 127, in queryset_chunks
    items = list(qs[i * chunk_size:][:chunk_size])
  File "/var/praekelt/python/local/lib/python2.7/site-packages/django/db/models/query.py", line 162, in __iter__
    self._fetch_all()
  File "/var/praekelt/python/local/lib/python2.7/site-packages/django/db/models/query.py", line 967, in _fetch_all
    self._prefetch_related_objects()
  File "/var/praekelt/python/local/lib/python2.7/site-packages/django/db/models/query.py", line 591, in _prefetch_related_objects
    prefetch_related_objects(self._result_cache, self._prefetch_related_lookups)
  File "/var/praekelt/python/local/lib/python2.7/site-packages/django/db/models/query.py", line 1516, in prefetch_related_objects
    obj_list, additional_lookups = prefetch_one_level(obj_list, prefetcher, lookup, level)
  File "/var/praekelt/python/local/lib/python2.7/site-packages/django/db/models/query.py", line 1615, in prefetch_one_level
    prefetcher.get_prefetch_queryset(instances, lookup.get_current_queryset(level)))
  File "/var/praekelt/python/local/lib/python2.7/site-packages/taggit/managers.py", line 137, in get_prefetch_queryset
    qs = self.get_queryset(query).using(db).extra(
TypeError: get_query_set() takes exactly 1 argument (2 given)
```

Running `print type(self)` just before the break gives:
`<class 'modelcluster.contrib.taggit._ClusterTaggableManager'>`

The cause seems to be a call to `_TaggableManager.get_prefetch_queryset()` which calls `self.get_query_set(query)`. While `_TaggableManager.get_query_set()` accepts an optional parameter, the the overridden implementation in `_ClusterTaggableManager` doesn't.

**Comment by gasman** [comment]
Hi @KaitCrawford - what versions of Wagtail, Django, django-modelcluster and django-taggit are you running?

**Comment by KaitCrawford** [comment]
Hi @gasman we're seeing the error on an environment running Wagtail 1.3.1, Django 1.8.9 (reproduced on 1.8.12), django-modelcluster 1.1 and django-taggit 0.18.0 (reproduced on 0.18.1)

**Comment by gasman** [comment]
Thanks! Have replicated on latest master now, by editing `BlogPage.search_fields` on wagtaildemo to:

    search_fields = Page.search_fields + (
        index.SearchField('body'),
        index.RelatedFields('tags', [
            index.SearchField('name', partial_match=True, boost=10),
        ]),
    )

**Comment by gasman** [comment]
Fixed in #55, although still with some room for improvement...

This occurs whenever a `prefetch_related` is performed on the tags field (which happens whenever you have it listed in search_fields) - and isn't specific to the latest versions of Django / taggit as I originally suspected, so I'm quite surprised that this hasn't come up before now!

The fix in #55 prevents the error, but doesn't actually do the prefetching, so the `update_index` operation won't be as efficient as we'd like - but at least this way it doesn't break completely.

**Comment by KaitCrawford** [comment]
@gasman, I see the fix for this has been added to v2 which hasn't been released yet. It's quite urgent to us to get it deployed. Is there a release date in the near future planned for this? If not, would it be possible to release the fix as a patch for an earlier version?

**Comment by gasman** [comment]
@KaitCrawford Yep - hoping to get a new release out today or Monday.

**Comment by KaitCrawford** [comment]
That's excellent news @gasman! thanks!

**Comment by gasman** [comment]
Have now released django-modelcluster 2.0, incorporating this fix.

**Comment by miltontony** [comment]
Thanks @gasman much appreciated!
